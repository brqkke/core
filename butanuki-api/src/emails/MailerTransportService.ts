import { Injectable } from '@nestjs/common';
import { createTransport, Transporter } from 'nodemailer';
import { AppConfigService } from '../config/app.config.service';

@Injectable()
export class MailerTransportService {
  private transporter: Transporter;

  constructor(private appConfig: AppConfigService) {
    this.transporter = this.makeMailTransport();
  }

  makeMailTransport() {
    return createTransport({
      maxConnections: 5,
      pool: true,
      host: this.appConfig.config.email.smtp.host,
      port: this.appConfig.config.email.smtp.port,
      secure: true,
      auth: {
        user: this.appConfig.config.email.smtp.username,
        pass: this.appConfig.config.email.smtp.password,
      },
    });
  }

  public async sendRawEmail({
    to,
    subject,
    content,
    prefix,
  }: {
    to: string;
    subject: string;
    content: string;
    prefix?: string;
  }) {
    //TODO move to queue
    const fullSubject =
      prefix === undefined ? `Butanuki - ${subject}` : `${prefix}${subject}`;
    const transport = this.transporter;
    return transport.sendMail({
      from: { name: 'Butanuki', address: 'no-reply@butanuki.com' },
      to,
      subject: fullSubject,
      html: content,
    });
  }
}
