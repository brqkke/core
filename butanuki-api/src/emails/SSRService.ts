import { Injectable } from '@nestjs/common';
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { EmailEnveloppe } from './templates/EmailEnveloppe';
import { I18nService } from '../i18n/i18n.service';

@Injectable()
export class SSRService {
  constructor(private i18nService: I18nService) {}

  render<Props extends object>(
    Element: React.FC<Props>,
    props: Props,
    locale: string,
  ) {
    return renderToStaticMarkup(
      React.createElement(
        EmailEnveloppe,
        {
          initialI18nStore: this.i18nService.getInitialI18nStore(),
          initialLanguage: locale,
        },
        React.createElement(Element, props),
      ),
    );
  }
}
