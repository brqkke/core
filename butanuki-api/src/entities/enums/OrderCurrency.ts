import { registerEnumType } from '@nestjs/graphql';

export enum OrderCurrency {
  CHF = 'CHF',
  EUR = 'EUR',
}

registerEnumType(OrderCurrency, { name: 'OrderCurrency' });
