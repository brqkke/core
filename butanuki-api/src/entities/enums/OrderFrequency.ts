import { registerEnumType } from '@nestjs/graphql';

export enum OrderFrequency {
  UNIQUE = 'UNIQUE',
  WEEKLY = 'WEEKLY',
  MONTHLY = 'MONTHLY',
}

registerEnumType(OrderFrequency, { name: 'OrderFrequency' });
