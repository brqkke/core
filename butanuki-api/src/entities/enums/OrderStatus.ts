import { registerEnumType } from '@nestjs/graphql';

export enum OrderStatus {
  TO_CANCEL = 'TO_CANCEL', //not used
  OPEN = 'OPEN',
  FILLED = 'FILLED',
  FILLED_NEED_RENEW = 'FILLED_NEED_RENEW',
  CANCELLED_NEED_RENEW = 'CANCELLED_NEED_RENEW', //not used
  CANCELLED = 'CANCELLED',
}

registerEnumType(OrderStatus, { name: 'OrderStatus' });
