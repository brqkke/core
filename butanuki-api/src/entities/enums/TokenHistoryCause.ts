export enum TokenHistoryCause {
  /** @deprecated */
  REFRESH = 'REFRESH',
  /** @deprecated */
  REFRESH_FROM_HISTORY = 'REFRESH_FROM_HISTORY',
  MANUAL_CHANGE = 'MANUAL_CHANGE',
}
