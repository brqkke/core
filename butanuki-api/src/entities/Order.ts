import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from 'typeorm';
import { OrderStatus } from './enums/OrderStatus';
import { User } from './User';
import { OrderCurrency } from './enums/OrderCurrency';
import { Vault } from './Vault';
import { Field, ObjectType } from '@nestjs/graphql';

@ObjectType()
export class BityPaymentDetails {
  @Field({ nullable: true })
  iban?: string;

  @Field({ nullable: true })
  swift_bic?: string;

  @Field({ nullable: true })
  recipient?: string;

  @Field({ nullable: true })
  account_number?: string;

  @Field({ nullable: true })
  bank_code?: string;

  @Field({ nullable: true })
  bank_address?: string;
}

@Unique(['remoteId'])
@Entity()
@ObjectType()
export class Order {
  @PrimaryGeneratedColumn('uuid')
  @Field()
  id: string;

  @Column()
  remoteId: string;

  @Column()
  @Field()
  transferLabel: string;

  @Column({ type: 'enum', enum: OrderStatus })
  @Field()
  status: OrderStatus;

  @Column()
  userId: string;

  @Index()
  @JoinColumn()
  @ManyToOne(() => User, (user) => user.orders)
  user?: User;

  @Column()
  createdAt: Date;

  @Column()
  updatedAt: Date;

  @Column()
  lastCheckedAt: Date;

  @Column({ type: 'string', nullable: true })
  previousOrderId?: string | null;

  @ManyToOne(() => Order, (order) => order.renewedByOrders, { nullable: true })
  @JoinColumn()
  previousOrder?: Order;

  @OneToMany(() => Order, (order) => order.previousOrder)
  renewedByOrders?: Order[];

  @ManyToOne(() => Vault, { nullable: true })
  @Index()
  @JoinColumn()
  vault?: Vault | null;

  @Column({ nullable: true })
  @Field(() => String, { nullable: true })
  vaultId?: string | null;

  @Column()
  @Field()
  amount: number;

  @Column({ type: 'enum', enum: OrderCurrency })
  @Field(() => OrderCurrency)
  currency: OrderCurrency;

  @Column()
  bankDetails?: string;

  @Column()
  @Field(() => String, { nullable: true })
  redactedCryptoAddress?: string;
}
