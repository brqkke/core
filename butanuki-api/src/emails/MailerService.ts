import { Injectable } from '@nestjs/common';
import { SSRService } from './SSRService';
import { MailerTransportService } from './MailerTransportService';
import { LoginLinkEmailContent } from './templates/LoginLinkEmailContent';
import { AppConfigService } from '../config/app.config.service';
import React from 'react';
import { BityAccountWasUnlinked } from './templates/BityAccountWasUnlinked';
import { NewOrderEmailContent } from './templates/NewOrderEmailContent';
import { Order } from '../entities/Order';
import { Token } from '../entities/Token';
import { TokenStatusReporting } from './templates/TokenStatusReporting';
import { I18nService } from '../i18n/i18n.service';

@Injectable()
export class MailerService {
  constructor(
    private renderer: SSRService,
    private transport: MailerTransportService,
    private appConfig: AppConfigService,
    private i18n: I18nService,
  ) {}

  private renderAndSend<P extends {}>({
    to,
    subject,
    element,
    params,
    prefix,
    locale,
  }: {
    to: string;
    subject: string;
    prefix?: string;
    element: React.FC<P>;
    params: P;
    locale?: string;
  }) {
    const content = this.renderer.render(element, params, locale || 'fr');
    return this.transport.sendRawEmail({ to, content, subject, prefix });
  }

  sendSingleUseLoginLink(to: string, token: string) {
    return this.renderAndSend({
      to,
      subject: 'Here is your login link',
      element: LoginLinkEmailContent,
      params: {
        url: `${this.appConfig.config.baseUrl}/login/verify/${token}/${to}`,
      },
    });
  }

  sendBityRelink(to: string, locale: string) {
    return this.renderAndSend({
      to,
      subject: this.i18n.getLng(locale).t('email.bity_token_broken.subject'),
      element: BityAccountWasUnlinked,
      params: { appUrl: this.appConfig.config.baseUrl },
      locale,
    });
  }

  async sendNewOrderEmail(order: Order, email: string, locale: string) {
    const bankDetails = JSON.parse(order.bankDetails || '{}') as {
      iban: string;
      swift_bic: string;
      recipient: string;
      account_number: string;
      bank_code: string;
      bank_address: string;
    };
    const { t } = this.i18n.getLng(locale || 'fr');
    return this.renderAndSend({
      to: email,
      element: NewOrderEmailContent,
      params: {
        amount: order.amount,
        currency: order.currency,
        redactedCryptoAddress: order.redactedCryptoAddress || '',
        reference: order.transferLabel,
        ...bankDetails,
      },
      subject: t('email.new_order.subject'),
      locale,
    });
  }

  async sendReportingEmail(content: string) {
    const now = new Date();
    return this.transport.sendRawEmail({
      to: this.appConfig.config.backgroundAgent.reporting.reportingEmail,
      content,
      subject: `Report ${now.getUTCFullYear()}-${now.getUTCMonth() + 1}`,
      prefix: 'BTNK ',
    });
  }

  async sendReportBityRefreshError(token: Token) {
    return this.renderAndSend({
      to: this.appConfig.config.backgroundAgent.reporting.alertEmail,
      params: { token },
      element: TokenStatusReporting,
      subject: `Bity token refresh error ${token.id}`,
      prefix: '[BTNK Alert] ',
    });
  }
}
