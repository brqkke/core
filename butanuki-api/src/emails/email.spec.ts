import { Order } from '../entities/Order';
import { OrderCurrency } from '../entities/enums/OrderCurrency';
import { OrderStatus } from '../entities/enums/OrderStatus';
import { MailerService } from './MailerService';
import { Test, TestingModule } from '@nestjs/testing';
import { MailerTransportService } from './MailerTransportService';
import { ModuleMocker, MockFunctionMetadata } from 'jest-mock';
import { SSRService } from './SSRService';

const moduleMocker = new ModuleMocker(global);

describe('Send new order email', () => {
  let mailerService: MailerService;
  let module: TestingModule;
  beforeEach(async () => {
    module = await Test.createTestingModule({
      providers: [MailerService],
    })
      .useMocker((token) => {
        if (token === MailerTransportService) {
          return { sendRawEmail: jest.fn() };
        }
        if (token === SSRService) {
          return new SSRService();
        }
        if (typeof token === 'function') {
          const mockMetadata = moduleMocker.getMetadata(
            token,
          ) as MockFunctionMetadata<any, any>;
          const Mock = moduleMocker.generateFromMetadata(mockMetadata);
          return new Mock();
        }
      })
      .compile();
    mailerService = module.get(MailerService);
  });

  it('sends an email with the order label', async () => {
    const order: Order = {
      amount: 10,
      bankDetails: JSON.stringify({
        iban: '1245 1234 1234',
        swift_bic: 'ABCDEF',
        recipient: 'Bity co',
        account_number: '9876543210',
        bank_code: '1234',
        bank_address: '42 space street',
      }),
      redactedCryptoAddress: 'aaa...zzz',
      currency: OrderCurrency.EUR,
      id: '12345',
      createdAt: new Date(),
      updatedAt: new Date(),
      transferLabel: 'bity.com 123456',
      lastCheckedAt: new Date(0),
      status: OrderStatus.OPEN,
      previousOrderId: null,
      remoteId: '123-123-123',
      userId: '123456',
    };
    const transport = module.get(MailerTransportService);
    await mailerService.sendNewOrderEmail(order, 'aa@aaa.com');
    expect(
      jest.mocked(transport.sendRawEmail).mock.calls[0][0],
    ).toMatchSnapshot();
  });
});
