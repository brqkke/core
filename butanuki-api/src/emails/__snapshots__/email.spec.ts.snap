// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email rendering send an email for bity relink 1`] = `
Object {
  "content": "<html><body><h1 style=\\"text-align:center\\">Butanuki</h1><h2>Your Bity account was unlinked</h2><p>You need to link it again in your Butanuki account if you want to keep your DCA working.</p><a href=\\"https://localhost:3000\\">Login to Butanuki</a></body></html>",
  "prefix": undefined,
  "subject": "It looks like your bity account is no more linked !",
  "to": "aa@aaa.com",
}
`;

exports[`Email rendering send an email for bity relink 2`] = `
Object {
  "content": "<html><body><h1 style=\\"text-align:center\\">Butanuki</h1><h2>Votre compte bity s&#x27;est délié</h2><p>Vous devez le lier à nouveau dans votre compte Butanuki si vous souhaitez que votre épargne récurrente continue de fonctionner.</p><a href=\\"https://localhost:3000\\">Se connecter à Butanuki</a></body></html>",
  "prefix": undefined,
  "subject": "Votre compte bity ne semble plus lié !",
  "to": "aa@aaa.com",
}
`;

exports[`Email rendering sends an email with the order label 1`] = `
Object {
  "content": "<html><body><h1 style=\\"text-align:center\\">Butanuki</h1><h2>New order</h2><p>Here are the details of your new order of <b>EUR 10</b> to aaa...zzz:</p><p>IBAN : <b>1245 1234 1234</b><br/>BIC/SWIFT : <b>ABCDEF</b><br/>Recipient : <b>Bity co</b><br/>Account number : <b>9876543210</b><br/>Bank Code : <b>1234</b><br/>Bank Address : <b>42 space street</b><br/>Reference : <b>bity.com 123456</b><br/></p></body></html>",
  "prefix": undefined,
  "subject": "New Butanuki order",
  "to": "aa@aaa.com",
}
`;

exports[`Email rendering sends an email with the order label 2`] = `
Object {
  "content": "<html><body><h1 style=\\"text-align:center\\">Butanuki</h1><h2>Nouvel ordre</h2><p>Voici les détails de votre nouvel ordre de <b>EUR 10</b> vers aaa...zzz :</p><p>IBAN : <b>1245 1234 1234</b><br/>BIC/SWIFT : <b>ABCDEF</b><br/>Recipient : <b>Bity co</b><br/>Account number : <b>9876543210</b><br/>Bank Code : <b>1234</b><br/>Bank Address : <b>42 space street</b><br/>Reference : <b>bity.com 123456</b><br/></p></body></html>",
  "prefix": undefined,
  "subject": "Nouvel ordre Butanuki",
  "to": "aa@aaa.com",
}
`;
